# Capstone 1

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns

import numpy as np
import pandas as pd

# Step 1: Load the Dataset

In [2]:
df=pd.read_csv(r"C:\Users\AMAR\Desktop\data scince modules 1 excel\Capstone 1\university ranking data (uncleaned).csv")
df

,Rank,institution,location code,location,ar score,ar rank,er score,er rank,fsr score,fsr rank,...,cpf rank,ifr score,ifr rank,isr score,isr rank,irn score,irn rank,ger score,ger rank,score scaled
0,1,Massachusetts Institute of Technology (MIT),US,United States,100.0,5,100.0,4,100.0,14,...,5,100.0,54,90.0,109,96.1,58,100.0,3,100.0
1,2,University of Cambridge,UK,United Kingdom,100.0,2,100.0,2,100.0,11,...,55,100.0,60,96.3,70,99.5,6,100.0,9,98.8
2,3,Stanford University,US,United States,100.0,4,100.0,5,100.0,6,...,9,99.8,74,60.3,235,96.3,55,100.0,2,98.5
3,4,University of Oxford,UK,United Kingdom,100.0,3,100.0,3,100.0,8,...,64,98.8,101,98.4,54,99.9,3,100.0,7,98.4
4,5,Harvard University,US,United States,100.0,1,100.0,1,99.4,35,...,2,76.9,228,66.9,212,100.0,1,100.0,1,97.6
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1417,1418,Universidad Nacional Autónoma de Honduras (UNAH),HN,Honduras,4.6,501+,5.8,501+,3.6,601+,...,601+,1.7,601+,NaN,NaN,8.1,601+,NaN,NaN,-
1418,1419,Universidad Tecnológica de Bolívar,CO,Colombia,3.0,501+,5.8,501+,2.4,601+,...,601+,9.3,601+,1.3,601+,8.8,601+,30.9,403,-
1419,1420,Université Mohammed V de Rabat,MA,Morocco,2.8,501+,2.6,501+,3.2,601+,...,601+,5.2,601+,2.8,601+,51.1,601+,NaN,NaN,-
1420,1421,University of Craiova,RO,Romania,3.3,501+,1.8,501+,2.7,601+,...,601+,NaN,NaN,NaN,NaN,7.5,601+,8.7,601+,-


In [3]:
df.columns 

Index(['Rank', 'institution', 'location code', 'location', 'ar score',
       'ar rank', 'er score', 'er rank', 'fsr score', 'fsr rank', 'cpf score',
       'cpf rank', 'ifr score', 'ifr rank', 'isr score', 'isr rank',
       'irn score', 'irn rank', 'ger score', 'ger rank', 'score scaled'],
      dtype='object')

In [4]:
df = df.rename(columns={
    'Rank': 'World Rank',
    'institution': 'University',
    'location code': 'Country Code',
    'location': 'Country',

    'ar score': 'Academic Reputation Score',
    'ar rank': 'Academic Reputation Rank',

    'er score': 'Employer Reputation Score',
    'er rank': 'Employer Reputation Rank',

    'fsr score': 'Faculty Student Ratio Score',
    'fsr rank': 'Faculty Student Ratio Rank',

    'cpf score': 'Citations per Faculty Score',
    'cpf rank': 'Citations per Faculty Rank',

    'ifr score': 'International Faculty Ratio Score',
    'ifr rank': 'International Faculty Ratio Rank',

    'isr score': 'International Student Ratio Score',
    'isr rank': 'International Student Ratio Rank',

    'irn score': 'International Research Network Score',
    'irn rank': 'International Research Network Rank',

    'ger score': 'Employment Outcomes Score',
    'ger rank': 'Employment Outcomes Rank',

    'score scaled': 'Overall Score'
})


In [5]:
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1422 entries, 0 to 1421
Data columns (total 21 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   World Rank                            1422 non-null   int64  
 1   University                            1422 non-null   object 
 2   Country Code                          1422 non-null   object 
 3   Country                               1422 non-null   object 
 4   Academic Reputation Score             1422 non-null   float64
 5   Academic Reputation Rank              1422 non-null   object 
 6   Employer Reputation Score             1421 non-null   float64
 7   Employer Reputation Rank              1422 non-null   object 
 8   Faculty Student Ratio Score           1420 non-null   float64
 9   Faculty Student Ratio Rank            1422 non-null   object 
 10  Citations per Faculty Score           1417 non-null   float64
 11  Citations per Fac

### Replace '-' With NaN

In [10]:
df.replace('-', np.nan, inplace=True)

In [11]:
df.isnull().sum().sort_values(ascending=False)


Overall Score                           922
International Faculty Ratio Score        98
International Faculty Ratio Rank         98
International Student Ratio Rank         57
International Student Ratio Score        57
International Research Network Rank      13
International Research Network Score     13
Employment Outcomes Rank                 12
Employment Outcomes Score                12
Citations per Faculty Score               5
Faculty Student Ratio Score               2
Employer Reputation Score                 1
World Rank                                0
Academic Reputation Score                 0
Country                                   0
Country Code                              0
University                                0
Citations per Faculty Rank                0
Faculty Student Ratio Rank                0
Academic Reputation Rank                  0
Employer Reputation Rank                  0
dtype: int64

# Drop Columns 
#### score scaled column shows 922 blank value

In [12]:
# Drop scored scaled column 

df.drop(columns=['Overall Score'], inplace=True)


# Handle Missing Values

### For Score Columns → optional median fill

In [13]:
score_cols = [col for col in df.columns if 'Score' in col]

df[score_cols] = df[score_cols].fillna(df[score_cols].median())


### For Rank Columns

In this dataset, missing values (NaN) are present in several ranking-related columns such as Academic Reputation Rank, Employer Reputation Rank, and other indicator-specific ranks. These missing values do not indicate poor performance or a low ranking. Instead, they represent cases where a university was not evaluated or not ranked for a specific indicator.

Since ranking variables are ordinal by nature, imputing or replacing missing values with numerical placeholders (such as 0, maximum rank, or average rank) would introduce incorrect assumptions and distort the relative ordering of institutions. Similarly, removing records with missing rank values would lead to unnecessary data loss and bias country-level and institution-level analysis.

Therefore, missing values in ranking columns were intentionally retained as NaN to preserve the integrity and meaning of the data. During analysis and visualization, these NaN values are interpreted as “Not Ranked” and are handled at the presentation layer (e.g., tooltips, filters, or explanatory notes) rather than being altered in the underlying data.

This approach ensures:

Accurate statistical calculations

Correct handling of numeric data types

Transparent and unbiased analytical insights

In [19]:
df.isnull().sum().sort_values(ascending=False)

International Faculty Ratio Rank        98
International Student Ratio Rank        57
International Research Network Rank     13
Employment Outcomes Rank                12
Country                                  0
Country Code                             0
University                               0
World Rank                               0
Academic Reputation Score                0
Academic Reputation Rank                 0
Employer Reputation Score                0
Employer Reputation Rank                 0
Citations per Faculty Rank               0
Citations per Faculty Score              0
Faculty Student Ratio Rank               0
Faculty Student Ratio Score              0
International Student Ratio Score        0
International Faculty Ratio Score        0
International Research Network Score     0
Employment Outcomes Score                0
dtype: int64

# Duplicate Universities check

In [15]:
df.duplicated(subset=['University', 'Country']).sum()


np.int64(0)

In [57]:
df.tail()

,Rank,institution,location code,location,ar score,ar rank,er score,er rank,fsr score,fsr rank,cpf score,cpf rank,ifr score,ifr rank,isr score,isr rank,irn score,irn rank,ger score,ger rank
1417,1418,Universidad Nacional Autónoma de Honduras (UNAH),HN,Honduras,4.6,501+,5.8,501+,3.6,601+,1.0,601+,1.7,601+,NaN,Unknown,8.1,601+,NaN,Unknown
1418,1419,Universidad Tecnológica de Bolívar,CO,Colombia,3.0,501+,5.8,501+,2.4,601+,2.2,601+,9.3,601+,1.3,601+,8.8,601+,30.9,403
1419,1420,Université Mohammed V de Rabat,MA,Morocco,2.8,501+,2.6,501+,3.2,601+,3.6,601+,5.2,601+,2.8,601+,51.1,601+,NaN,Unknown
1420,1421,University of Craiova,RO,Romania,3.3,501+,1.8,501+,2.7,601+,2.0,601+,NaN,Unknown,NaN,Unknown,7.5,601+,8.7,601+
1421,1422,University of Oradea,RO,Romania,4.0,501+,2.1,501+,3.3,601+,1.5,601+,1.7,601+,7.9,601+,16.6,601+,9.7,601+


# Checking Duplicates

In [16]:
df.duplicated().sum()


np.int64(0)

# overll conclusion
“I standardized column names, converted placeholder values (-) into proper missing values, validated data types, removed a highly incomplete Overall Score column to prevent biased insights, handled remaining missing values using median imputation, and finalized the dataset for Power BI visualization.”

# Save Clean Dataset

In [22]:
df.to_csv("university_ranking_cleaned_2026.csv", index=False)


In [23]:
import os
os.getcwd()

'C:\\Users\\AMAR'